ownCloud/Client/Actions/Scanner/ScanViewController.swift

```diff
@@ -0,0 +1,526 @@
+		let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "page", for: indexPath)
+
+		if let pageCell = cell as? ScanPageCell {
+			pageCell.aspectHeight = height - (verticalPadding * 2)
+			pageCell.page = pages[indexPath.item]
+		}
+
+		return cell
+	}
+}
+
+extension CGRect {
+	func asData() -> NSData {
+		return NSData(from: self)
+	}
+}
+
+struct ScanExportFormat {
+	var name : String
+	var suffix : String
+	var supportsMultiPage : Bool = false
+	var exporter : ((_ baseURL: URL, _ pages: [ScanPage]) -> ([URL]?))?
+}
+
+@available(iOS 13.0, *)
+class ScanViewController: StaticTableViewController {
+	weak var core : OCCore?
+	var targetFolderItem : OCItem?
+
+	// Pages
+	var pagesSection : StaticTableViewSection
+	var pagesCollectionViewController : ScanPagesCollectionViewController?
+
+	// Save
+	var saveSection : StaticTableViewSection
+	var fileNameRow : StaticTableViewRow?
+
+	// Options
+	var optionsSection : StaticTableViewSection
+	var formatSegmentedControl : UISegmentedControl?
+	var oneFilePerPageRow : StaticTableViewRow?
+
+	private static func exportAsImage(format: String, baseURL: URL, pages: [ScanPage]) -> [URL]? {
+		var files : [URL] = []
+
+		for page in pages {
+			let targetURL = baseURL.appendingPathComponent(UUID().uuidString)
+			var imageData : Data?
+
+			switch format {
+				case "jpeg":
+					imageData = page.image.jpegData(compressionQuality: 0.75)
+
+				case "png":
+					imageData = page.image.pngData()
+
+				default: break
+			}
+
+			if let imageData = imageData {
+				if (try? imageData.write(to: targetURL)) != nil {
+					// Writing successful
+					files.append(targetURL)
+				} else {
+					// Error writing
+					return nil
+				}
+			} else {
+				// Error
+				return nil
+			}
+		}
+
+		return files.count > 0 ? files : nil
+	}
+
+	private static func exportAsPDF(pages: [ScanPage], to url: URL) -> Bool {
+		if let pdfContext = CGContext(url as CFURL, mediaBox: nil, nil) {
+			defer {
+				pdfContext.closePDF()
+			}
+
+			for page in pages {
+				if let jpegData = page.image.jpegData(compressionQuality: 0.75),
+				   let dataProvider = CGDataProvider(data: jpegData as CFData),
+				   let jpegImage = CGImage(jpegDataProviderSource: dataProvider, decode: nil, shouldInterpolate: true, intent: .defaultIntent) {
+					let mediaBoxRect = CGRect(x: 0, y: 0, width: page.image.size.width, height: page.image.size.height)
+
+					pdfContext.beginPDFPage([
+						kCGPDFContextMediaBox : mediaBoxRect.asData()
+					] as CFDictionary)
+
+					pdfContext.draw(jpegImage, in: mediaBoxRect)
+
+					pdfContext.endPDFPage()
+				} else {
+					return false
+				}
+			}
+
+			return true
+		}
+
+		return false
+	}
+
+	let availableExportFormats : [ScanExportFormat] = [
+		ScanExportFormat(name: "PDF", suffix: "pdf", supportsMultiPage: true, exporter: { (baseURL, pages) in
+			let targetURL = baseURL.appendingPathComponent(UUID().uuidString)
```

Comment:
@felix-schwarz Any reason using special URL containing random UUID just for PDFs? AFAIK UUID is quite computationally intensive.

Reply:
The exporter for every format can be called multiple times (f.ex. when exporting one PDF file for every scanned page of a multi-page document). This is an easy way to ensure there won't be any name collisions.

Regarding computational intensity: UUIDs are specified as having to be cryptographically secure, so yes. But they're used so widely these days (including in the SDK) and their creation so fast that I'd assume modern CPUs have a special instruction for their creation. #interestingResearchTopic :-D

However, the number of UUIDs used here is really small, so there shouldn't be any noticeable impact on battery life or performance due to this.